modules/bidViewability.js

```diff
@@ -0,0 +1,88 @@
+// This module, when included, will trigger a BID_VIEWABLE event which can be consumed by Bidders and Analytics adapters
+// GPT API is used to find when a bid is viewable, https://developers.google.com/publisher-tag/reference#googletag.events.impressionviewableevent
+// Does not work with other than GPT integration
+
+import { config } from '../src/config.js';
+import * as events from '../src/events.js';
+import { EVENTS } from '../src/constants.json';
+import { logWarn, isFn, triggerPixel } from '../src/utils.js';
+import { getGlobal } from '../src/prebidGlobal.js';
+import adapterManager, { gdprDataHandler, uspDataHandler } from '../src/adapterManager.js';
+import find from 'core-js-pure/features/array/find.js';
+
+const MODULE_NAME = 'bidViewability';
+const CONFIG_FIRE_PIXELS = 'firePixels';
+const CONFIG_CUSTOM_MATCH = 'customMatchFunction';
+const BID_VURL_ARRAY = 'vurls';
+const GPT_IMPRESSION_VIEWABLE_EVENT = 'impressionViewable';
+
+export let isBidAdUnitCodeMatchingSlot = (bid, slot) => {
+  return (slot.getAdUnitPath() === bid.adUnitCode || slot.getSlotElementId() === bid.adUnitCode);
+}
+
+export let getMatchingWinningBidForGPTSlot = (globalModuleConfig, slot) => {
+  return find(getGlobal().getAllWinningBids(),
+    // supports custom match function from config
+    bid => isFn(globalModuleConfig[CONFIG_CUSTOM_MATCH])
+      ? globalModuleConfig[CONFIG_CUSTOM_MATCH](bid, slot)
+      : isBidAdUnitCodeMatchingSlot(bid, slot)
+  ) || null;
+};
+
+export let fireViewabilityPixels = (globalModuleConfig, bid) => {
+  if (globalModuleConfig[CONFIG_FIRE_PIXELS] === true && bid.hasOwnProperty(BID_VURL_ARRAY)) {
+    let queryParams = {};
+
+    const gdprConsent = gdprDataHandler.getConsentData();
+    if (gdprConsent) {
+      if (typeof gdprConsent.gdprApplies === 'boolean') { queryParams.gdpr = Number(gdprConsent.gdprApplies); }
+      if (gdprConsent.consentString) { queryParams.gdpr_consent = gdprConsent.consentString; }
+      if (gdprConsent.addtlConsent) { queryParams.addtl_consent = gdprConsent.addtlConsent; }
+    }
+
+    const uspConsent = uspDataHandler.getConsentData();
+    if (uspConsent) { queryParams.us_privacy = uspConsent; }
+
+    bid[BID_VURL_ARRAY].forEach(url => {
+      // we are assuming that "?" will be already present in the url
```

Comment:
Is assuming '?' at the end of urls a good assumption?

Reply:
The assumption is made as these pixels will have additional params (for bid details) already present in URL, and then we need to add consent params. 
Most of the ad-tech companies use standard way of adding params in URL ( with `?` and `&` ), if there is going to be an exception then we will not be sure what delimiters are being used in such case.

Reply:
I have removed this assumption now, now the code will add the '?' if it is not present in URL